src/hotspot/share/runtime/os.cpp

```diff
@@ -1857,3 +1845,82 @@ void os::naked_sleep(jlong millis) {
   }
   naked_short_sleep(millis);
 }
+
+
+////// Implementation of pagesizeset_t
+
+// A pagesizeset_t is a set containing a set of page sizes.
+
+// sets the given page size
+void os::PagesizeSet::add(size_t pagesize) {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  _v |= pagesize;
+}
+
+// returns true if given page size is part of the set
+bool os::PagesizeSet::is_set(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  return _v & pagesize;
+}
+
+// returns the next smallest page size set in this set, or 0.
+size_t os::PagesizeSet::next_smaller(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  // mask out all pages sizes >= pagesize:
+  uintx v2 = _v & pagesize - 1;
+  if (v2 > 0) {
+    return round_down_power_of_2(v2);
+  }
+  return 0;
+}
+
+size_t os::PagesizeSet::next_larger(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  if (pagesize == max_power_of_2<uintx>()) { // Shift by 32/64 would be UB
+    return 0;
+  }
+  int l = exact_log2(pagesize) + 1;
+  uintx v2 = _v >> l;
+  if (v2 == 0) {
+    return 0;
+  }
+  while ((v2 & 1) == 0) {
+    v2 >>= 1;
+    l ++;
+  }
+  return (size_t)1 << l;
+}
+
+size_t os::PagesizeSet::largest() const {
+  const size_t max = max_power_of_2<uintx>();
+  if (is_set(max)) {
+    return max;
+  }
+  return next_smaller(max);
+}
+
+size_t os::PagesizeSet::smallest() const {
+  assert(min_page_size() > 0, "Sanity");
+  return next_larger(min_page_size() / 2);
+}
+
+void os::PagesizeSet::print_on(outputStream* st) const {
+  bool first = true;
+  for (size_t sz = smallest(); sz != 0; sz = next_larger(sz)) {
+    if (first) {
+      first = false;
+    } else {
+      st->print_raw(", ");
+    }
+    if (sz < M) {
+      st->print(SIZE_FORMAT "k", sz / K);
+    } else if (sz < G) {
+      st->print(SIZE_FORMAT "m", sz / M);
+    } else {
```

Comment:
The prefixes for mega/gigabyte are "M" and "G" instead of "m" and "g" respectively